src/platform/utilities/data/debounce.js

```diff
@@ -0,0 +1,13 @@
+/**
+ * @param {function}
+ * @param {number}
+ */
+export default function debounce(wait, func) {
+  let timeoutId;
+
+  return function debouncedFunction(...args) {
+    const context = this;
+    clearTimeout(timeoutId);
+    timeoutId = setTimeout(() => func.apply(context, args), wait);
```

Comment:
Can we avoid using the `this` workaround here?

Reply:
I don't think so. I took this pretty much straight from https://medium.com/@griffinmichl/implementing-debounce-in-javascript-eab51a12311e when I realized I was going to have some `this` context issues.

Reply:
Thanks for the `context`, it's not clear to me what assigning this to context is accomplishing? Could you explain?

Reply:
Yeah, absolutely! No guarantees it'll be a _good_ explanation, but here goes:

First, we start off by returning a [closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) `debouncedFunction`. This will modify the `timeoutId`. Note the use of the `function` keyword here instead of `=>`. What this means is when an object calls the `debouncedFunction`, the `this` context will be the calling object and _not_ `debounce`. That's what we want.

The trouble comes in when we use `setTimeout`. When `setTimeout` calls the function it's passed,  it's a "free function invocation" (I don't know if that's a standard term, but the article I linked to used it), which means the lexical `this` binding is more or less tossed out the window. The function `setTimeout` calls has `this` attached to nothing.

To overcome that, we use [`function.prototype.apply`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply) to give it the `this` we want.

Finally, `setTimeout`'s first parameter takes a function, so we can't just call `setTimeout(func.apply(context, args), wait)`. We need to pass it a function. We use an arrow function here because the `this` binding of the function we pass it doesn't _really_ matter; when it's called, it'll run the function we want with the right context now.

Reply:
Thanks @cvalarida, using `apply` and a traditional anonymous function (non-arrow) solves the global scope issue with `setTimeout`, I tried logging `this` in the `setTimeout` call with the workaround (storing `this` in the `context `variable): 
```
timeoutId = setTimeout(() => {console.log('this', context); func.apply(context, args);}, wait);
```
and without:
```
timeoutId = setTimeout(() => {console.log('this', this); func.apply(this, args);}, wait);
```
and it printed `AutosuggestField` regardless (in fact, in this case, the method is already bound before we pass it into debounce, so we could pass `null` as `this` in `apply`, and it still works). 

We may also want to .cancel() the `debouncedGetOptions` , so that it does not attempt to change state after the component is unmounted.

Reply:
Yeah, good point about `.cancel()`. I'll go ahead and add that.